### DISCUSSION TITLE
Can I use Server Component with a global Context?

### DISCUSSION THREAD

**Discussion by hucancode** [vercel/next.js#42301] [App Router]
I am seeing the whole point of Server Component is that it gets rendered on the server. But what if I want to implement i18n, or theme context, or anything that requires having a global context.
Take this `app/layout.jsx` for example, I want to have an `I18n` context available site-wide, so I can translate whatever text whenever I want.
```html
    <html>
      <body className="page-container">
        <I18n lang={lang} dictionary={dictionary}>
          <Navbar />
          <main>{children}</main>
          <FootNote />
        </I18n>
      </body>
    </html>
```
The `I18n` component is implemented as so, using `createContext`, I am forcing myself to abandon Server Component.
```js
"use client";
import { createContext, useContext } from "react";
import rosetta from "rosetta";
export const I18nContext = createContext();
const i18n = rosetta();
export default function I18n({ lang, dictionary, children }) {
  i18n.set(lang, dictionary);
  i18n.locale(lang);
  return <I18nContext.Provider value={i18n}>{children}</I18nContext.Provider>;
}

export function useI18n() {
  return useContext(I18nContext);
}
```
As what I understand, this approach forces almost everything to be client component, child of a client component has to be client component, right?
Do we have any solution to implement global context but still render on server whenever possible?

**Comment by krall12** [comment]
As I understand it what you're doing should theoretically work. You can have client components that take server components as children. See https://github.com/reactjs/rfcs/blob/main/text/0188-server-components.md#capabilities--constraints-of-server-and-client-components

**Reply by oliviertassinari** [reply]
But does it mean that these server components are rendered as client components? So have their JavaScript shipped to the browser?

**Reply by felixhaeberle** [reply]
This would interest me too!!!

**Reply by yakhinvadim** [reply]
> But does it mean that these server components are rendered as client components? So have their JavaScript shipped to the browser?

Doesn't seem like it. Docs say:
> React will know it needs to render <ServerComponent /> on the server before sending the result (which doesn't contain any server-only code) to the client. From the Client Component's perspective, its child will be already rendered.

https://beta.nextjs.org/docs/rendering/server-and-client-components#importing-server-components-into-client-components

**Reply by markomitranic** [reply]
UPDATE: I was previously spreading misinformation that I've once heard from Vercel employees, among others. I have performed a test to settle the misunderstanding. I have run the test on local dev, local prod build and vercel prod build.

This is a screenshot of what HTML the browser receives as part of the initial document, with JS disabled. The data inside of the blue component is stuff that I fetch from a local JSON file on the disk - it is a server component.

![Screenshot 2024-01-19 at 12 29 39](https://github.com/vercel/next.js/assets/10811624/c5611874-6f19-493d-93c9-df83ccd4cb40)

To take things further:
If we look at the HTML of the document we will see a bunch of embeded JS/JSON content near the bottom. We can easily identify that the render result corresponding to the server component is stored here as JSON, and will be used during re-rendering of the outer client component. Yes, not the code of the server component, instead a static result of the code that was executed on the server.

![Screenshot 2024-01-19 at 13 17 29](https://github.com/vercel/next.js/assets/10811624/d1711408-2a7b-448b-b5a4-7391ce41b139)

**Comment by hucancode** [comment]
Nice. Thank you! @krall12

**Comment by amannn** [comment]
I asked the same question in regards to i18n this week on Twitter: https://twitter.com/jamannnnnn/status/1587835293989732352.

It seems to me like having to switch to client components to be able to use context is very much at odds with enabling as many components as possible to stay on the server side.

I'm currently looking for such a capability in order to [bring RSC support to `next-intl`](https://github.com/amannn/next-intl/pull/139), an i18n library I'm maintaining.

**Comment by leerob** [comment]
We now have documentation on Context: https://beta.nextjs.org/docs/rendering/server-and-client-components#context

**Reply by amannn** [reply]
Hey Lee, thanks for mentioning this here! 

I have to say though that the i18n use case still seems a bit hard to achieve with the presented guidelines.

The main issue I'm seeing is that with the currently available options, this forces many components to be marked for the client side, which could otherwise remain server-only. The docs mention that React Context is typically used for data that changes over time, but I think for the i18n use case this is not true.

~~Have you considered one of these approaches?~~
1. ~~`createServerContext` currently creates a context object that is automatically passed to the client-side as far as I can tell. Have you considered changing it to do the inverse, i.e. only being available on the server side? Based on the naming that's what I thought it might do. For auto-passing context to the client, maybe there could be `createClientContext` or just the existing `createContext`?~~
2. ~~A request-level object, like `context` in GraphQL resolvers. This is not hierarchically available like React Context, but instead something that can be instantiated once per request and then be available everywhere. E.g. initialization in a middleware and then having a function like `context()` (similar to `headers()`) would be ideal IMO.~~



~~**EDIT**: The [currently proposed `createServerContext` API](https://twitter.com/sebmarkbage/status/1587615489605369861?lang=en) would work for the i18n use case (based on the limited information that's currently available), by only moving the locale there and then using it to resolve other i18n config via `React.cache`. Would be great if `createServerContext` is supported in Next.js!~~

**EDIT 2:** Reading params deeply would solve this problem without context: https://github.com/vercel/next.js/discussions/58862

Thank you! 🙂

**Reply by amannn** [reply]
~@leerob Any chance we could get access to [`requestAsyncStorage`](https://github.com/vercel/next.js/blob/canary/packages/next/client/components/request-async-storage.ts) via a public API? This would work 😬: https://github.com/amannn/next-intl/blob/9365fe614a557eff3bbfa67c1d170bcab4125826/packages/example/src/utils/Storage.tsx~

**Reply by raphaelbadia** [reply]
Getting access to requestAsyncStorage would be great, I'd like to have it too 😕

**Reply by markomitranic** [reply]
@amannn ~what did you end up using? I’m looking at a use case similar to yours where the value isnmt expected to change and I purely wish to prevent prop drilling for locale string.~

ouch. so they didnt communicate even with someone like you lol

> How does unstable_setRequestLocale work?
> next-intl uses [cache()](https://react.dev/reference/react/cache) to create a mutable store that holds the current locale. By calling unstable_setRequestLocale, the current locale will be written to the store, making it available to all APIs that require the locale.

**Reply by amannn** [reply]
@markomitranic Yep, next-intl currently uses a workaround. I hope that some form of this will land: https://github.com/vercel/next.js/discussions/58862

**Comment by manvalls** [comment]
I've made a small utility for this using `cache`:
https://www.npmjs.com/package/server-only-context

**Reply by jdgamble555** [reply]
I copied your idea and made a custom hook - https://dev.to/jdgamble555/easy-context-in-react-server-components-rsc-1mdf

**Comment by unleashit** [comment]
I also think it'd be nice to see a per request server context made available in some form. I'm new to Next and was pretty baffled that the docs discount the need. There are plenty of uses beyond i18n. For example, I tried to build a serverside caching singleton (the suggested fetch/cache solution or prop drilling is simple for my needs) but ran up against a wall trying to keep it alive from beginning to end of a request (and not persist between requests).

**Reply by manvalls** [reply]
See my answer above: https://www.npmjs.com/package/server-only-context

**Reply by unleashit** [reply]
Hey I did check that out, but the way it's set up in the example won't anything set to serverContext() persist between requests and possibly leak unwanted data? My singleton works in a similar way, but with the idea of cleaning its state on each request. The problem is when I try to do the cleaning in the root layout.ts, it sometimes gets run multiple times during the request (thus cleaning the cache too soon).

In a related issue, I've noticed that while layout.ts is the first component to be called,  React continues to render children (page.ts, etc.) before any promises inside layout resolve. Since I'd like some root data to be avail regardless of route, and I prefer to call it in a single place it seems I'm out of luck. Would be nice to have an entry point where you could temporarily block rendering if needed. If I understand this new paradigm correctly, I guess the idea is to discourage this in order to let async streaming do its work. Still best to err on the side of flexibility IMO.

**Reply by manvalls** [reply]
That's not the case @unleashit as it uses `cache` under the hood, which means the set value is only valid during the request, and it works even when handling several requests in parallel at the same time, each one having its own independent value.

**Reply by unleashit** [reply]
Thanks for the reply. If that's the case, then I imagine  it must work because React's cache() has access to the request lifecycle behind the scenes. It would be nice if that access wasn't black boxed and we were provided a way to hook into it ourselves. Not sure if it's planned, or if they've just determined it a foot gun without a valid use.

**Comment by leerob** [comment]
Hello! We have some documentation here on recommended patterns for using React Context: https://beta.nextjs.org/docs/rendering/server-and-client-components#context

**Reply by amannn** [reply]
Hey @leerob!

Is there any roadmap or similar to follow along with the development of `createServerContext`? Seems like [it should work in Next.js](https://twitter.com/dan_abramov/status/1641176794941509632) at some point?

I do understand that you'd want to use regular React Context for most cases, but seems like there are [valid cases for `createServerContext`](https://twitter.com/sebmarkbage/status/1587615489605369861?lang=en).

**Reply by unleashit** [reply]
Hey thanks for chiming in. I've seen those docs, but I guess some of us would like to do things that are not necessarily the recommended pattern. I get that Server Components can have a couple of different contexts: 1. as part of an initial full page load and 2. a later update from the client (as far as I know, the future plan is to allow the component to be individually called?). So I can see the problem of having code be aware more than one context. In addition, new things that get added to context could be ignored once streaming begins (which could requiring some blocking if those things are important). Still, I would much prefer easy access to req and res (and/or the Node Async Context) and accept the complexity tradeoff.

While it would be easier to access context (on the server directly), I found I was able to achieve what I wanted with React.Cache. At least in my testing so far, I'm able to cache a singleton instance (redux store, etc.) and have it be replaced at new requests as I'd expect. Are there other hidden gotchas besides the ones I mentioned to why you don't seem to want to support this?

**Reply by einarq** [reply]
@leerob Does those docs have a new home? That link doesn't work anymore.

**Comment by MythicManiac** [comment]
This became a blocker for us just recently. We're aiming to build an architecture where our standalone component library only depends on a generic data provider interface without binding to any specific implementation. Each environment where the component library is then used (including but not limited to NextJS projects) would provide their own implementation of the data provider. A similar need exists when it comes e.g. to theming or even the link component implementation (given that NextJS apps benefit from NextJS links, but other apps might use something else).

Originally we set out to use context for this, but as others already noted it won't work on the server.

Here's a workaround that seemed to work so far, although it is somewhat cursed and might lead to rehydration issues down the line if we're not careful, as we now can provide a different data provider instance on the client & server (which could lead to different rendering results between the two): https://github.com/thunderstore-io/thunderstore-ui/pull/744

I don't particularly like having to implement a workaround in the first place, it'd be great if NextJS would support something to enable the kind of pluggable architecture we're looking to implement.

**Comment by ajwootto** [comment]
Ran into this as well when trying to build a feature flagging SDK, which requires a shared "context" for the duration of the request that can hold the current user and flag configuration data.

The solution I ended up landing on was basically to "curry" the whole interface of the SDK via a setup function that defines how to obtain the user for the current request. You can see that in action [here](https://docs.devcycle.com/sdk/client-side-sdks/nextjs/nextjs-usage-app#create-the-devcycle-context-and-export-it) and view the source code implementation [here](https://github.com/DevCycleHQ/js-sdks/blob/main/sdk/nextjs/src/server/setupDevCycle.ts#L17). It basically looks like this

```typescript
// export "curried" functions that are scoped to use the options passed here
export const { getVariableValue } = setupDevCycle({
   sdkKey: 'some-key',
   userGetter: async () => {
      const userIdentity = headers().get('user-id')
      return {
         user_id: userIdentity
      }
   }
})
```

The reason this works is that all the methods that are exported from that setup function end up calling an "ensure initialization" function which performs things like fetching the current configuration and current user data according to the configured options. It then memoizes the results using the React `cache` API in a manner similar to some of the comments above. That means that all subsequent calls to the functions will just use the memoized result for the rest of this request's lifecycle.

Now the interface within a server component becomes:
```typescript
// import the "curried" function from the file where the setup function is called
import { getVariableValue } from './devcycle'

export default async ServerComponent() {
  // this function can be called with just the details of the flag we want to evaluate. It already knows the current user's data and
  // what the flagging configuration is
  const variableValue = await getVariableValue('someKey', false)
  
  // do something with the value
}
```

Hopefully this helps someone, I was stuck on this for a while.

**Reply by markomitranic** [reply]
Great example of using a singleton as a cache initializer. You didn't have to curry it tho, that part was just syntax sugar.

**Reply by ajwootto** [reply]
It was a requirement of the interface we wanted for the SDK to have these functions be "context aware", otherwise you would need to provide the user's data on each call. 

The reason for that is that originally we thought we could create a single "setup" location that would be called from a single place in the rendering tree (ideally the root layout) and would then ensure that all subsequent calls to access the cache from elsewhere in the tree would be able to retrieve that data from setup. 

Something like this:
```typescript
// app/layout.ts

export default async function RootLayout({
  children,
}: {
  children: React.ReactNode;
}) {
  const user = await getUser();

  setupDevCycle({user, sdkKey: 'some-key'})

  return (
    <html>
      <body>
          {children}
      </body>
    </html>
  );
}
```

and then all subsequent children of that root layout would have access to the user's data via the shared cache. 

However that approach didn't work because
1. Layouts aren't rendered before pages, they seem to be rendered after
2. Layouts are cached on clientside navigation so their side effects don't run 

So you can't rely on the "setup" having run before the children (page etc.) are rendered. Unless I missed something, there didn't seem to be a place in the rendering tree where you could "guarantee" that something had run before subsequent things were rendered, except perhaps in the "page" itself. That would mean that each page using our SDK would need to call the setup function with the same options and SDK key and the same way of determining the user's data, which we deemed too cumbersome.

So instead we landed on this function that implicitly performs the setup and populates the cache no matter where it's called from, otherwise it uses what's already in cache.

**Reply by unleashit** [reply]
Yep. The problem I ran into with the singleton idea is there is no "entry file" you can count on. This is also a problem if you need to set something up like Sentry that needs their middlewares to be placed at specific points in the request. On the server, Next's root layout is called after the route's page.tsx . So if you call something like setupDevCycle() in a layout, your context won't be available (at least deterministically) everywhere you might need it. My "solution" was to initialize it in every page.tsx, This has worked for me on a simple statically exported site, but it's not well tested as at least for now I've now given up on Next.js unless/until it becomes more flexible and less opinionated.

**Reply by ajwootto** [reply]
Yeah I also tried an approach where the `getVariableValue` function would sit and check every tick for the context to become available, which occurred whenever the layout got around to being rendered. That seemed to work and solve problem 1 above but it seemed to add a 5-15 ms delay to each page render because it was basically "de-optimizing" whatever Next was trying to do by rendering the page before the layout. Then it still didn't work for problem 2 because the layout is only _definitely_ executed on the initial page request, and is usually cached on clientside nav.

**Reply by markomitranic** [reply]
@unleashit  that is the only currently accepted solution afaik. This is also how ‘next-intl’ works, you have to invoke a cache based singleton in each layout AND each page (where you wish to use it).

Does the answer above resolve that in any way for you? Am i misunderstanding it?

**Reply by unleashit** [reply]
Interesting. Whether this theory is true, it would be so nice if these types of inner workings were documented wouldn't it? I never did perf tests (to notice such small delays), but I did end up with a request safe context that seemed to work pretty well. Forgot to mention above that I also wrapped the instance in a React.cache(). Here's an experimental package I made (that no one should probably use since I made it before server actions were even a thing): https://github.com/unleashit/next-promise-cache 

Edit: @markomitranic I replied before I saw your latest. I don't like it as a solution because a) it has to be repeated everywhere and b) it would be much better if Next.js provided the developer with access to the same Node.js async local storage that they are utilizing.

**Reply by ajwootto** [reply]
Something that would solve this is a "before rendering" callback you can define that is guaranteed to always be called to perform any initial setup, tracer initialization etc. 

I know they don't want to give us the full request object for various reasons, but even just having a defined place to hook into the lifecycle while still using the existing request-wrapping APIs like `headers` and `cookies` would go a long way.

**Reply by ajwootto** [reply]
@markomitranic from my understanding the reason that `next-intl` needs to work that way is because it needs access to the current page's `props` in order to determine what locale to use. The difference with my usecase here is that I don't need anything from those props (that are only available from a `page` function) and everything that needs to be done per-request can be defined in a single place (the user getter above, which can do things like read from headers and cookies to determine who our user is)

**Reply by unleashit** [reply]
> I know they don't want to give us the full request object for various reasons

I get that this can be a foot gun, but maybe enable with a flag? This is the minimum I would need to look at next.js again personally.

**Reply by ajwootto** [reply]
I think it also has something to do with Next being able to use static analysis to determine whether a page can be statically rendered (ie. any page that doesn't import headers, cookies etc. is safe to be static).

**Reply by unleashit** [reply]
Could that be solved by allowing to opt out of automatic detection? If the dev doesn't have the settings right, it's their fault? I do get that's it's complicated (and amazing) to provide all the feature's Next provides. That said, using it felt like 5 steps forward and 10 steps back after using other React SSR frameworks (that granted ONLY did SSR, but didn't get in the way constantly with strong opinions).

**Reply by ajwootto** [reply]
I think another reason they don't allow it is that not every request to the backend is the same "type" of request. On clientside nav the browser is sending a request that's basically asking for a partial update of the server-rendered content. Those details are abstracted behind the helpers like `headers` that give you safe ways to interact with the parts of the request that are always available.

On the other hand there definitely seems to be an ongoing debate about "performance by default" vs. "expose more details and let developers do it". One other example is the current monkeypatching the framework does on the `fetch` function to cache everything by default even when not wanted.

**Comment by don-esteban** [comment]
Exactly today I struggled with a similar challenge. After only studying the docs (@leerob linked to them) I was even more confused. The docs are excellent in my opinion and they get better almost every day. Forgive me, if I missed the point, I did not read the whole conversation, but I'll focus now on the initial post.

@hucancode said: 

> ...As what I understand, this approach forces almost everything to be client component, child of a client component has to be client component, right?...

Right, but to "switch back" to React Server Components, you have this "slot pattern"

[Supported Pattern: Passing Server Components to Client Components as Props](https://nextjs.org/docs/app/building-your-application/rendering/composition-patterns#supported-pattern-passing-server-components-to-client-components-as-props),

which is not helpful at the first view. To use React Context you place a custom client context provider in your RootLayout, or somewhere close to the component root. But now everything enclosed in this lets say `<ThemeProvider>` are client components?

Yes and no. This is not true for PageComponents `page.jsx`. You can use Server Components inside Client Components (slot pattern) but you can't import them. You do not import page components. I'll guess this handles the framework and uses the slot pattern under the hood. But React/Next people know much better. I'm still in the beginning.

```
// RootLayout
export default function RootLayout({ children }: { children: React.ReactNode }) {
  return (
    <html>
      <ThemeProvider value={theme}>
          <body>{children}</body>
      </ThemeProvider>
    </html>
  );
}
```

So you still can continue with nested pages as Server Components and access the context in even deeper nested client components.

But what, if you want to access the Context in a Server Component? You use this "slot pattern" and forward the context values as props to the React Server Component.

This was a "quicky", so hopefully did not oversee things.

```
// A page component
"use server"; // default

import ClientComponent from "./client-component";
import ServerComponent from "./server-component";

export default async function PageComponent() {
  return (
    <ClientComponent>
      <ServerComponent />
    </ClientComponent>
  );
}
```

```
// A client component with access to Context (probably, I did not test this)
import React, { ReactElement } from 'react';

export default async function ClientComponent({children}: {children: ReactElement}) {

  const serverProp = "The client pulled this value out of context"
  
  return (
    <div>
      {React.cloneElement(children, { serverProp })}
    </div>
  );
}
```

```
// Server Component using Context
import React, { ReactElement } from 'react';

export default async function ServerComponent({serverProp}: {serverProp?: string}) {
  
  return (
    <div>
      I am a server component. I got this prop from the client: {serverProp}
    </div>
  );
}

```

@leerob You have this very helpfull "Good to know" sections in the docs. If It is not just me, who was a bit confused by the fact that Client Components can not import Server components, perhaps it's worth adding a hint.

**Reply by markomitranic** [reply]
Huh this makes no sense to me. The whole point of server components is that they get rendered on the server, and are passed to client components as props, as rendered. This means that applying props on those rendered children client-side through React.cloneElement won’t do anything as they are rendered and don’t contain the code needed to update their state or print the new value - it will just clone the static html of the server component as-is.

**Reply by don-esteban** [reply]
Yes, I see the point. Tomorrow I'll rethink. There must be a simple solution. I think, this is a common use case and there must be a simple pattern for those things.

**Reply by don-esteban** [reply]
**No, I don't see the point anymore :-)** Kind of tired! Tomorrow more, just this as a quick feedback.

The client gets back rendered content and (!) client side code.

```
// Server Component
import React from 'react';
import MyButton from './button';

export default async function ServerComponent({serverProp}: {serverProp?: string}) {
  
  return (
    <div>
      I am a server component. I got this prop from the client: {serverProp}
      <MyButton value={"I am at the client now again: " +  serverProp} />
    </div>
  );
}

// Wrap button as a client component.

"use client"

import React from 'react';

export default function MyButton({value}: {value: string}) {
  
  return (
    <button onClick={() => alert(value)}>Click me</button>
  );
}
```

**Reply by markomitranic** [reply]
Afaik the only currently accepted solution os usage of react’s cache. This is what all the libraries that use server context, like ‘next-intl’ use.

A user above has created an utility to make this slightly easier https://www.npmjs.com/package/server-only-context

**Reply by don-esteban** [reply]
Thanks for the link @markomitranic. That's interesting. But I don't see how this helps in the context of "switching between or combining" Client Components and Server Components. Perhaps we are talking about different things?

Part of the initial question I reference to:

> As what I understand, this approach forces almost everything to be client component, child of a client component has to be client component, right?
> Do we have any solution to implement global context but still render on server whenever possible?

**Reply by markomitranic** [reply]
I understand the confusion, the first question was a bit too broad and asked 2 questions in one without knowing it :) 

> this approach forces almost everything to be client component, child of a client component has to be client component, right?

This is not true, as shown in previous comments and in the Next.js docs. A server component can be rendered on the server, or statically, and served as a child of a client component when passed as params. 

This leaves us only with the second question:

> Do we have any solution to implement global context but still render on server whenever possible?

This is what I was responding to - on the client you simply use a context to have a global context - on the server you use the cache trick in all your layouts and pages, which gives you the same functionality, the ability to set data globally during a request and share it between components.

A good example of this is `next-intl` which (unless you wish to prop-drill the locale string into every single component on both server and client) uses the context on the client, and `cache` on the server to share global state between components. In fact there is a npm library to make this server part easier -  https://www.npmjs.com/package/server-only-context

To answer a third question, that nobody asked yet, but will - no, it is not possible to share context data between client and server. You can set it individually, or share it unidirectionally from server to the client, but not the other way around.

Edit: formatting fix

**Reply by ajwootto** [reply]
I think the way to think of it is that if you want a context shared between the whole tree (server and client) then it's the server that needs to own that context, not the client. In the [docs for my library](https://docs.devcycle.com/sdk/client-side-sdks/nextjs/nextjs-usage-app#create-the-devcycle-context-and-export-it) I linked in the above comment you can see how the serverside context setup returns a function `getClientContext` which is then passed to a `ClientsideProvider` component. That way the server components can access the context data directly from the function by importing it, while the client components can receive the same shared data via their usual "context" mechanism and the `useContext` hook.

**Reply by LikeDreamwalker** [reply]
Thanks and I finally undertsand this part. In this article [Supported Pattern: Passing Server Components to Client Components as Props](https://nextjs.org/docs/app/building-your-application/rendering/composition-patterns#supported-pattern-passing-server-components-to-client-components-as-props), it declare something:

1. We **can't** import a server component into a client component
2. We **can** use `children` prop to send the server component into a client component

For the conclusion 2, the server component will still works as the server-rendered component, so **you don't need to worry about that**. And I really can't remember where does this idea come from, that a server component wrapped by client component will be rendered  as client component anyway, maybe the original idea is comes from the conclusion 1.

And for the conclusion 1, this is normal if you have build a server component, but you use it in a parent client component, which can be like a menu or list that needs React. At this time you should receive an issue.

So for this discussion, **if you have built some providers, like zustand, next-themes or next-intl and so on, it will be normal that they declare that they are client components, _but they will not influence your children server components in layout_.** 

And we can do a simple test, in my code, I use zustand, mui, shadcn/ui and next-intl and so on:
```tsx
    <html suppressHydrationWarning>
        <CommonStoreProvider>
          <AppRouterCacheProvider options={{ key: "css" }}>
            <NextIntlClientProvider messages={messages}>
              <ThemeProvider
                attribute="class"
                defaultTheme="system"
                enableSystem
              >
                <MuiThemeProvider theme={theme} defaultMode="system">
                  <CssBaseline />
                  {children}
                  <LayoutClient />
                </MuiThemeProvider>
              </ThemeProvider>
            </NextIntlClientProvider>
          </AppRouterCacheProvider>
        </CommonStoreProvider>
      </body>
    </html>
```
And I didn't decalre `'use client'` in the layout, also in my page:
```tsx
    <>
      <LandingMainArea locale={locale}></LandingMainArea>
      <LPClient locale={locale}></LPClient>
    </>
```
`LPClient` is a client component, and `LandingMainArea` is a server component, which works normal as a server-rendered component. And I keep the rules to avoid using React or any other features that RSC doesn't support. But what if we add it on?

```tsx
export default async function LandingMainArea({ locale }: Props) {
  // RSC solution from next-intl
  const t = await getTranslations("Common");
  // zustand store provider which works under react context
  const {
    globalExtendPopupOption,
    setGlobalExtendPopupOption,
    goExtend,
    setClientReady,
    goFeatureCheck,
  } = useCommonStore((store) => ({
    globalExtendPopupOption: store.globalExtendPopupOption,
    setGlobalExtendPopupOption: store.setGlobalExtendPopupOption,
    goExtend: store.goExtend,
    setClientReady: store.setClientReady,
    goFeatureCheck: store.goFeatureCheck,
  }));
  return (
...
```
And you will see this issue definitely:
![image](https://github.com/user-attachments/assets/6624b3d4-8a4d-4a72-857c-e54f9d5c50a3)
This means the `LandingMainArea` component works.

_In case you get confused again later on, if you are not sure where your server component is working, you can just add some client-side codes to see if they will cause errors._

**Reply by LikeDreamwalker** [reply]
> Thanks and I finally undertsand this part. In this article [Supported Pattern: Passing Server Components to Client Components as Props](https://nextjs.org/docs/app/building-your-application/rendering/composition-patterns#supported-pattern-passing-server-components-to-client-components-as-props), it declare something:
> 
> 1. We **can't** import a server component into a client component
> 2. We **can** use `children` prop to send the server component into a client component
> 
> For the conclusion 2, the server component will still works as the server-rendered component, so **you don't need to worry about that**. And I really can't remember where does this idea come from, that a server component wrapped by client component will be rendered as client component anyway, maybe the original idea is comes from the conclusion 1.
> 
> And for the conclusion 1, this is normal if you have build a server component, but you use it in a parent client component, which can be like a menu or list that needs React. At this time you should receive an issue.
> 
> So for this discussion, **if you have built some providers, like zustand, next-themes or next-intl and so on, it will be normal that they declare that they are client components, _but they will not influence your children server components in layout_.**
> 
> And we can do a simple test, in my code, I use zustand, mui, shadcn/ui and next-intl and so on:
> 
> ```tsx
>     <html suppressHydrationWarning>
>         <CommonStoreProvider>
>           <AppRouterCacheProvider options={{ key: "css" }}>
>             <NextIntlClientProvider messages={messages}>
>               <ThemeProvider
>                 attribute="class"
>                 defaultTheme="system"
>                 enableSystem
>               >
>                 <MuiThemeProvider theme={theme} defaultMode="system">
>                   <CssBaseline />
>                   {children}
>                   <LayoutClient />
>                 </MuiThemeProvider>
>               </ThemeProvider>
>             </NextIntlClientProvider>
>           </AppRouterCacheProvider>
>         </CommonStoreProvider>
>       </body>
>     </html>
> ```
> 
> And I didn't decalre `'use client'` in the layout, also in my page:
> 
> ```tsx
>     <>
>       <LandingMainArea locale={locale}></LandingMainArea>
>       <LPClient locale={locale}></LPClient>
>     </>
> ```
> 
> `LPClient` is a client component, and `LandingMainArea` is a server component, which works normal as a server-rendered component. And I keep the rules to avoid using React or any other features that RSC doesn't support. But what if we add it on?
> 
> ```tsx
> export default async function LandingMainArea({ locale }: Props) {
>   // RSC solution from next-intl
>   const t = await getTranslations("Common");
>   // zustand store provider which works under react context
>   const {
>     globalExtendPopupOption,
>     setGlobalExtendPopupOption,
>     goExtend,
>     setClientReady,
>     goFeatureCheck,
>   } = useCommonStore((store) => ({
>     globalExtendPopupOption: store.globalExtendPopupOption,
>     setGlobalExtendPopupOption: store.setGlobalExtendPopupOption,
>     goExtend: store.goExtend,
>     setClientReady: store.setClientReady,
>     goFeatureCheck: store.goFeatureCheck,
>   }));
>   return (
> ...
> ```
> 
> And you will see this issue definitely: ![image](https://private-user-images.githubusercontent.com/45194556/396761289-6624b3d4-8a4d-4a72-857c-e54f9d5c50a3.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MzQ0OTczNzAsIm5iZiI6MTczNDQ5NzA3MCwicGF0aCI6Ii80NTE5NDU1Ni8zOTY3NjEyODktNjYyNGIzZDQtOGE0ZC00YTcyLTg1N2MtZTU0ZjlkNWM1MGEzLnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNDEyMTglMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjQxMjE4VDA0NDQzMFomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPTM2YTczY2Q1Mzk2ZWUyYjVhZTNmMGM5ZjQ0ZWZlZmRjNzYzYjQyZmFhMzYyZmFmMDJmYjVjMzczN2YzMDY3ZjgmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0In0.tOngXoDR0cJ6ymsrzHzT3G0dC9QdfDKIS6iaXdGYWj4) This means the `LandingMainArea` component works.
> 
> _In case you get confused again later on, if you are not sure where your server component is working, you can just add some client-side codes to see if they will cause errors._

And I think for the client component import RSC part, if your RSC is easy and not using some features from RSC (like async/await), you won't receive the error and Next.js will downgrade it to the normal client component. That's should be where the idea comes.

**Reply by benzsuankularb** [reply]
Still losing some benefits of having server-side rendering.
Cuz the server component in children won't get rendered at first load.
